### PR TITLE
Reduce complexity of Clowder methods

### PR DESCRIPTION
**Pull request by adamrdrew**
The goal of this PR is to reduce the complexity of Clowder methods as much as possible during the tech debt sprint. In an effort to go about this in a data driven way I'm using complexity measures based on cyclonic complexity and halstead metrics as provided by the go-complexity-analysis go vet plugin. The goal to to get as many high complexity methods under a cyclonic measure of 10 as possible within the time frame.

### CODE REVIEW

**Comment by adamrdrew** [comment]
Gave up on keda provider method `getTriggerRoute` - I tried to refactor it multiple times and there's something wrong with my approach. I don't see the problem no matter how much I look and I can't reproduce the failures locally. Decided to just give up on it becuase the method has a high complexity score, but it is artificial - there's a huge empty switch block in it.

**Comment by adamrdrew** [comment]
OK, so its passing tests now. I'll do some tidying up and then get it ready for review.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+func (r *ClowdAppReconciliation) steps() []func() (ctrl.Result, error) {
+	return []func() (ctrl.Result, error){
+		r.getApp,
+		r.setPresentAndManagedApps,
+		r.startMetrics,
+		r.isAppMarkedForDeletion,
+		r.addFinalizer,
+		r.isEnvLocked,
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:63]: I wonder if we could/should rename some of these a little and combine them. Right now there are a lot of steps to visually go through, for example, what about
`getClowdEnv` also chain calling, `isEnvCoreNamespaceDeleted`, `isClowdEnvReady`, and `isClowdEnvReconciled`

Perhaps it needs a different name, but I think that would collapse things down a little in this list of steps.

It could also be argued that the same could be done for `getApp`.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) stopMetrics() (ctrl.Result, error) {
+	r.reconciliationMetrics.stop()
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) setPresentAndManagedApps() (ctrl.Result, error) {
+	if _, ok := presentApps[r.app.GetIdent()]; !ok {
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:106]: In looking at this - I think we can simplify this - I know it's copy pasta, but why not

``presentApps[r.app.GetIdent()] = true``

can replace the if block

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+		presentApps[r.app.GetIdent()] = true
+	}
+	presentAppsMetric.Set(float64(len(presentApps)))
+
+	delete(managedApps, r.app.GetIdent())
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:111]: Can you add a comment here - it took me a moment to grok what I did - we remove it from the managed list because it _may_ have been managed before, but it may not be after this reconcile.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+		if k8serr.IsNotFound(getAppErr) {
+			// Must have been deleted
+			return ctrl.Result{}, errors.New(SKIPRECONCILE)
+		}
+		r.log.Info("App not found", "env", r.app.Spec.EnvName, "app", r.app.GetIdent(), "err", getAppErr)
+		return ctrl.Result{}, getAppErr
+	}
+	//This is kinda side-effecty but I couldn't think of a better place
+	//to put it.
+	logWithEnv := r.log.WithValues("env", r.app.Spec.EnvName)
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:127]: This is a side effect of the steps. Piece, one could argue it could become its own step. It also doesn't matter when it happens, so we could move this little piece out and into the actual reconciliation function as it references the EnvName from the spec and not from a real ClowdEnv, which is only gotten half way through the steps.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+	r.log = &logWithEnv
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) isAppMarkedForDeletion() (ctrl.Result, error) {
+	isAppMarkedForDeletion := r.app.GetDeletionTimestamp() != nil
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:134]: As soon as we get this in, I want to migrate this out and into the rhc-osdk-utils - No action, just an FYI

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+			r.log.Info("Set status error", "err", setClowdStatusErr)
+			return ctrl.Result{Requeue: true}, setClowdStatusErr
+		}
+		return ctrl.Result{}, getEnvErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) isEnvCoreNamespaceDeleted() (ctrl.Result, error) {
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:223]: Do we need `Core` in here?

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+	}
+
+	if ens.ObjectMeta.DeletionTimestamp != nil {
+		r.log.Info("Env target namespace is to be deleted - skipping reconcile")
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdAppReconciliation) isAppCoreNamespaceDeleted() (ctrl.Result, error) {
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:236]: Same as above

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 6 above)
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:223]: These two functions are identical apart from the name, should we create a combined function for this?

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+
+func (r *ClowdAppReconciliation) isEnvReady() (ctrl.Result, error) {
+	if !r.env.IsReady() {
+		r.recorder.Eventf(r.app, "Warning", "ClowdEnvNotReady", "Clowder Environment [%s] is not ready", r.app.Spec.EnvName)
+		r.log.Info("Env not yet ready")
+		if setClowdStatusErr := SetClowdAppConditions(*r.ctx, r.client, r.app, crd.ReconciliationFailed, fmt.Errorf("clowd env not ready")); setClowdStatusErr != nil {
+			r.log.Info("Set status error", "err", setClowdStatusErr)
+			return ctrl.Result{Requeue: true}, setClowdStatusErr
+		}
+
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_reconciliation.go:271]: blank line remove

File: controllers/cloud.redhat.com/clowdenvironment_controller.go [hunk 9]
```diff
@@ -107,237 +107,70 @@ func ReadEnv() string {
...
-//Reconcile fn
-func (r *ClowdEnvironmentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+//Acts as a guard for a reconciliation cycle, as well as initatizes a bunch of required objects
+func (r *ClowdEnvironmentReconciler) initReconciliation(ctx context.Context, req ctrl.Request) (logr.Logger, context.Context, crd.ClowdEnvironment, ctrl.Result, error) {
 	log := r.Log.WithValues("env", req.Name).WithValues("rid", utils.RandString(5))
 	ctx = context.WithValue(ctx, errors.ClowdKey("log"), &log)
 	ctx = context.WithValue(ctx, errors.ClowdKey("recorder"), &r.Recorder)
 	env := crd.ClowdEnvironment{}
 
+	silentFailError := errors.New(SILENTFAIL)
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_controller.go:117]: Why pass this back at all? Why not just pass back nil? Then there is no need to do the conditional later in the `Reconcile` function

File: controllers/cloud.redhat.com/clowdenvironment_controller.go [hunk 10]
```diff
@@ -107,237 +107,70 @@ func ReadEnv() string {
...
 
 	delete(managedEnvironments, env.Name)
+}
 
-	defer func() {
-		managedEnvsMetric.Set(float64(len(managedEnvironments)))
-	}()
+//Reconcile fn
+func (r *ClowdEnvironmentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log, ctx, env, res, err := r.initReconciliation(ctx, req)
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_controller.go:142]: A few things around this
1) I'm not sure I understand why we are doing it differently to the clowdapp
2) Passing in the context and getting a modified version back seems rather strange to me
I propose making it like ClowdApp

File: controllers/cloud.redhat.com/clowdenvironment_controller.go [hunk 11]
```diff
@@ -107,237 +107,70 @@ func ReadEnv() string {
...
-		log.Info("GetEnvResourceStatus error", "err", getEnvResErr)
-		return ctrl.Result{Requeue: true}, getEnvResErr
-	}
-
-	envStatus := core.ConditionFalse
-	successCond := cond.Get(&env, crd.ReconciliationSuccessful)
-	if successCond != nil {
-		envStatus = successCond.Status
-	}
+	r.initMetrics(env)
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_controller.go:154]: Is there a reason that metrics is handled differently here?

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 12]
```diff
@@ -0,0 +1,489 @@
...
+
+//During reconciliation we handle errors in 2 ways: sometimes we want to error out of reconciliation and sometimes we want to skip reconciliation.
+func shouldSkipReconciliation(err error) bool {
+	if err != nil && err.Error() == SKIPRECONCILE {
+		return true
+	}
+	return false
+}
+
+func MakeClowdEnvReconciliation(cache *rc.ObjectCache,
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:36]: Same question as for ClowdApp

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+	cache    *rc.ObjectCache
+	recorder record.EventRecorder
+	ctx      *context.Context
+	client   client.Client
+	env      *crd.ClowdEnvironment
+	log      *logr.Logger
+}
+
+//Returns a list of step methods that should be run during reconciliation
+func (r *ClowdEnvironmentReconciliation) steps() []func() (ctrl.Result, error) {
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:66]: In thinking about the steps piece here - the only thing I'm concerned about is that if we need to inject something in the middle it's a whole new step - I guess that's OK, yeh I guess that's fine - sorry stream of consciousness

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+		r.deleteUnusedResources,
+		r.setClowdEnvConditions,
+		r.logSuccess,
+	}
+}
+
+//Public method to iterate through the steps of the reconciliation process
+func (r *ClowdEnvironmentReconciliation) Reconcile() (ctrl.Result, error) {
+	r.log.Info("Reconciliation started")
+	//The env stays locked for the entire reconciliation
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:90]: Not a problem with this change

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 15]
```diff
@@ -0,0 +1,489 @@
...
+		result, err := step()
+		if err != nil {
+			return result, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+//Determine if env is marked for deletion, and if so finalize and end resonciliation
+func (r *ClowdEnvironmentReconciliation) markedForDeletion() (ctrl.Result, error) {
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:105]: Comment as in app

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 16]
```diff
@@ -0,0 +1,489 @@
...
+func (r *ClowdEnvironmentReconciliation) initTargetNamespace() (ctrl.Result, error) {
+	result := ctrl.Result{}
+	var err error = nil
+
+	if r.env.Status.TargetNamespace != "" {
+		return result, err
+	}
+
+	if r.env.Spec.TargetNamespace == "" {
+		result, err = r._makeTargetNamespace()
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:199]: Why the underscores?

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 17]
```diff
@@ -0,0 +1,489 @@
...
+	if ens.ObjectMeta.DeletionTimestamp != nil {
+		r.log.Info("Env target namespace is to be deleted - skipping reconcile")
+		return ctrl.Result{}, errors.New(SKIPRECONCILE)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClowdEnvironmentReconciliation) runProviders() (ctrl.Result, error) {
+	provider := r._makeProvider()
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:279]: Is this just to reduce line count of the function? I'd perhaps argue in a case like this, that being able to see the literal would help understanding.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go [hunk 18]
```diff
@@ -0,0 +1,489 @@
...
+		}
+		return ctrl.Result{Requeue: true}, setAppErr
+	}
+	return ctrl.Result{}, nil
+}
+
+//Performs the actual business logic for the setAppInfo step
+//TODO this is a bear of a method. Factor this out into pieces.
+func (r *ClowdEnvironmentReconciliation) _setAppInfoImplementation() error {
+	p := r._makeProvider()
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:335]: You don't actually need a provider object now, you already have everything you need on `r` :)

**Review by psav — CHANGES_REQUESTED** [review]
First pass review - love love love :heart: the work in here, if has made things ~im~measurably better, a few questions though and a few things that I think are bugs

File: controllers/cloud.redhat.com/providers/deployment/impl.go [hunk 19]
```diff
@@ -152,45 +138,95 @@ func initDeployment(app *crd.ClowdApp, env *crd.ClowdEnvironment, d *apps.Deploy
 		if readinessProbe.FailureThreshold == 0 {
 			readinessProbe.FailureThreshold = 3
 		}
-	} else if bool(deployment.Web) || deployment.WebServices.Public.Enabled {
-		readinessProbe = baseProbe
+
+		c.ReadinessProbe = &readinessProbe
+	}
+	if bool(deployment.Web) || deployment.WebServices.Public.Enabled {
```
[the remark this answers is not in the document]
Comment (adamrdrew) [reply controllers/cloud.redhat.com/providers/deployment/impl.go:144]: Good call! Changed

File: controllers/cloud.redhat.com/clowdapp_controller.go [hunk 20]
```diff
@@ -138,192 +133,24 @@ func (r *ClowdAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
-
-	cacheErr := cache.ApplyAll()
+	log.Info("Reconciliation started")
 
-	if cacheErr != nil {
-		r.Recorder.Eventf(&app, "Warning", "FailedReconciliation", "Clowdapp requeued [%s]", app.GetClowdName())
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, cacheErr); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
+	reconciliation := MakeClowdAppReconciliation(ctx, &cache, r.Client, r.Recorder, &app, log, &req)
```
[the remark this answers is not in the document]
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go:145]: Yeah I tried to avoid judgement calls where possible but this was one of them so I figured I'd try it and see what happens.

I don't have any kind of mental model for resource usage and performance impacts re: Clowder running in real life. Other than "try to use as few resources as possible" I don't have any sort of heuristics around that in my head about like what is and is not expensive relatively speaking. So in moving this code I didn't know if I was doing something as cheap as allocating an empty struct or as expensive as spinning up some complex system. Considering the comment I suspect its more like the latter than the former.

I'll change this to be a step method in `ClowdAppReconciliation` and I'll do it as late as possible in the process.

[the remark this answers is not in the document]
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go]: Moved it to a step!

File: controllers/cloud.redhat.com/clowdapp_controller.go [hunk 21]
```diff
@@ -138,192 +133,24 @@ func (r *ClowdAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+	res, err := reconciliation.Reconcile()
+	if err != nil {
+		if shouldSkipReconciliation(err) {
+			return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go]: Ah ok, it should be 

```
	res, err := reconciliation.Reconcile()
	if err != nil {
		if shouldSkipReconciliation(err) {
			return res, nil
		}
		return res, err
	}
```

Because some skips requeue:
```go
func (r *ClowdAppReconciliation) isEnvLocked() (ctrl.Result, error) {
	if ReadEnv() == r.app.Spec.EnvName {
		r.recorder.Eventf(r.app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", r.app.Spec.EnvName)
		r.log.Info("Env currently being reconciled")
		return ctrl.Result{Requeue: true}, errors.New(SKIPRECONCILE)
	}
	return ctrl.Result{}, nil
}
```

And some don't:
```go
func (r *ClowdAppReconciliation) isAppDisabled() (ctrl.Result, error) {
	if r.app.Spec.Disabled {
		r.log.Info("Reconciliation aborted - set to be disabled")
		return ctrl.Result{}, errors.New(SKIPRECONCILE)
	}
	return ctrl.Result{}, nil
}
```

Which matches the behavior of the old code:
```go

	if ReadEnv() == app.Spec.EnvName {
		r.Recorder.Eventf(&app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", app.Spec.EnvName)
		log.Info("Env currently being reconciled")
		return ctrl.Result{Requeue: true}, nil
	}
...
	if app.Spec.Disabled {
		log.Info("Reconciliation aborted - set to be disabled")
		return ctrl.Result{}, nil
	}
```

I changed it to reture the result - the steps were already doing the right thing but the skip eval wasn't respecting requeues

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go [hunk 22]
```diff
@@ -0,0 +1,381 @@
...
+	ctx                   *context.Context
+	client                client.Client
+	log                   *logr.Logger
+	app                   *crd.ClowdApp
+	req                   *ctrl.Request
+	env                   *crd.ClowdEnvironment
+	reconciliationMetrics ReconciliationMetrics
+}
+
+func MakeClowdAppReconciliation(ctx context.Context,
```
[the remark this answers is not in the document]
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:38]: Good catch re: public method. It is a hold over from an earlier design where this code was in its own package. Changed to private.

Its a method because you slapped my hand once for doing a literal declaration instead of using a constructor function :) There was probably a more subtle point to your asking me to use a function in the past, maybe like init logic or something, but it stuck in my head as "Pete wants me to use methods when creating instances of complex types. I can change it.

[the remark this answers is not in the document]
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:38]: Removed

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 1 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:61]: I would like to push back on this suggestion. My goal was for each step to (as much as possible) do a single thing and to have a complexity score of 1 or 2 - 3 max. This makes each step easy to reason about, and will make it very easy to write tests for each step. If the steps do more then there will be high conditionality, more ways to return, and will make testing more complex. (Note: I know tests aren't included in the patch. I ran out of time. But in future I'd like to add tests for every step.)

I do think this is the best way to bake this particular cake.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 2 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:105]: Good call. I had noticed that during the refactor and was like "Oh god, what am I not understanding about this block. It must be like this for a reason." I ended up leaving it out of fear lol. Changed.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 3 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:110]: Done

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 4 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:126]: Yeah this is a tough little one. Where I wanted to put this was at the very top of `ClowdAppReconciler.Reconcile` where we do all of those `r.Log.WithValues` calls, but I couldn't figure out how to get access to the env name before getting the app.  If there is a way to get the env name before getting the app then that would be a good change I think, but I don't know how.

So then the most strictly idiomatic thing would be to put it in a step, right? Yeah, it would be, except that this log update has no conditionality associated with it. It can't (so far as I can tell) fail. Additionally, there's no implementation to test, as this is a single public method on the logging library and thus outside of scope for testing. So, in this particular case a step would buy us nothing: no improvement to testing and  no reduction in complexity.

So long story short (too late) if you can tell me a way to grab env name before loading the app I'll move it out of the step. If not then I suggest we leave it as is.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 5 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:133]: Good call :+1:

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 6 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:222]: That was some cargo cult naming on my part. I didn't know if the fact that we were dealing with a core namespace was in some way significant. Judging from context I'm assuming it isn't :) I'll remove it..

Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go]: Oh yeah good call

Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go]: Consolidated them

File: controllers/cloud.redhat.com/clowdenvironment_controller.go (hunk 9 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_controller.go:117]: Yeah that's a little complex. I'll explain why it is that way and then let me know if you see a better way to handle it.

The core problem here is that the Reconcile function takes 3 different actions depending on the result of attempting to get the env from the client:
- If there's no error move on with the reconcile
- If the error is that the environment was deleted return an empty result and nil, terminating the reconcile but not returning an error
- If the error is any other error then terminate the reconcile and return the error

So the case of the env having been deleted is a special case where we silently fail. That's how it worked in the original code:

```go
if getEnvErr := r.Client.Get(ctx, req.NamespacedName, &env); getEnvErr != nil {
		if k8serr.IsNotFound(getEnvErr) {
			// Must have been deleted
			return ctrl.Result{}, nil
		}
		log.Info("Namespace not found", "err", getEnvErr)
		return ctrl.Result{}, getEnvErr
	}
```

So when I factored this out into a method I had to provide some way to tell the Reoncile function "Hey this is the special case where you want to bail based on an error but where you don't want to return the error. I didn't see any other way to handle this, and looking at the code I still don't (except to forgo using the special SILENTFAIL error I created and just move the `k8serr.IsNotFound` check into the error handler in Reconcile.

So open to suggestions on this one for sure.

File: controllers/cloud.redhat.com/clowdenvironment_controller.go (hunk 10 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_controller.go]: Done

File: controllers/cloud.redhat.com/clowdenvironment_controller.go (hunk 11 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_controller.go:150]: To try and match the original implementation as closely as possible. Metrics are initialized in the clowd environment far earlier in the Reconcile method than clowd app, before the point where the step system takes over. In clowd app the metrics are initialized in the middle of the process. I'm happy to put it wherever you suggest but this was me trying to preserve as much of the original logic flow as I could.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 12 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go]: Removed

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 8 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:247]: removed

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 13 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:50]: Yeah I would actually say that's a huge benefit. The reconcile function before was very hard to understand for someone new (or even not so knew to the project) - really had to take it in like a novel and understand it cover to cover. You had to read it and say "OK we're doing this here, and doing this here" etc. Now you can just read the step list and its like "Oh, ok, I get that." And if you want to add a step you know exactly where you are in the process, what's come before and will happen after, etc. Its maybe my favorite piece of this refactor. Seeing reconciliation as this list of steps I think is a bonus.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 14 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:74]: Awesome! I was super nervous about this.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 16 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:183]: Yeah, I wasn't sure how you'd react to that. I can change it if you don't like it and I'm not married to it. Basically, I wanted some way to denote what is and is not a step method. I could preface every step method with `step` but that would get ugly fast, and considering steps are 90% of the methods that didn't seem sensible. Scope can't help us because all these methods are private. If each step was its own type then each step's Run method would be public and any helper methods would be private and limited to that type. So, the best worst solution I landed on was to use an underscore to hint that this method is a private implementation detail and not a step.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 17 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:263]: No it was just being DRY. We create the provider in two places and if I notice something being done in 2 places I'll factor it out. None of the logic in the code deals with the internal state of the provider, we just pass it out somewhere so I figured this made sense.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 18 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:319]: I would request that we do that later because this method is a straight copy and paste from the original code and I don't understand what it is doing. Considering the size and complexity of the method I'd prefer to leave it as is (which is exactly what was in clowder before) and then we tackle it later.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 15 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go:89]: :+1:

File: controllers/cloud.redhat.com/providers/deployment/impl.go [hunk 23]
```diff
@@ -119,6 +93,11 @@ func initDeployment(app *crd.ClowdApp, env *crd.ClowdEnvironment, d *apps.Deploy
 		SuccessThreshold:    1,
 		TimeoutSeconds:      1,
 	}
+}
+
+func setLivenessProbe(pod crd.PodSpec, deployment crd.Deployment, env *crd.ClowdEnvironment, c *core.Container) {
```
[the remark this answers is not in the document]
Comment (psav) [reply controllers/cloud.redhat.com/providers/deployment/impl.go]: Discussed - resolving as we need to ensure Clowder politely corrects anything the user may have changed/altered.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 16 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go]: Renamed the underscore methods to funcNameImplementation in cases where funcName already existed as a step. Otherwise just gave them sane names.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 18 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go]: Refactored. You were right, way easier than I expected. Just deleted the provider and let the red squiggles tell me what to fix.

File: controllers/cloud.redhat.com/clowdenvironment_reconciliation.go (hunk 17 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_reconciliation.go]: Discussed. Removed function

File: controllers/cloud.redhat.com/clowdenvironment_controller.go (hunk 11 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdenvironment_controller.go:150]: Discussed. Leaving it as is.

File: controllers/cloud.redhat.com/clowdapp_controller.go [hunk 24]
```diff
@@ -138,192 +132,27 @@ func (r *ClowdAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+	reconciliation := ClowdAppReconciliation{
+		ctx:      &ctx,
+		client:   r.Client,
+		recorder: r.Recorder,
+		app:      &app,
+		log:      &log,
+		req:      &req}
+	res, err := reconciliation.Reconcile()
+	if err != nil {
+		if shouldSkipReconciliation(err) {
```
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_controller.go:150]: Sorry @adamrdrew I think there are some issues with this still. The main problem as far as I see is how the requeue is to be handled. In k8s, if you return a `nil` error from a reconciliation, the requeue request is ignored. It actually won't reconcile again ever, (until another requests comes in to reconcile). So this code returning `res` along with `nil` is actually equivalent to passing `ctrl.Result{}, nil`.

File: controllers/cloud.redhat.com/clowdapp_reconciliation.go (hunk 3 above)
Comment (psav) [reply controllers/cloud.redhat.com/clowdapp_reconciliation.go:93]: I don't see a comment :)

**Review by psav — CHANGES_REQUESTED** [review]
Tiny changes here I think

File: controllers/cloud.redhat.com/clowdapp_controller.go (hunk 24 above)
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]: Hmmm, is this a problem with the new implementation or what the problem always there?

This is the logic in the original code that I based this on:
```golang
	if ens.ObjectMeta.DeletionTimestamp != nil {
		log.Info("Env target namespace is to be deleted - skipping reconcile")
		return ctrl.Result{}, nil
	}
```

In our new step code that is recast as:
```golang
	if ens.ObjectMeta.DeletionTimestamp != nil {
		r.log.Info("Env target namespace is to be deleted - skipping reconcile")
		return ctrl.Result{}, errors.New(SKIPRECONCILE)
	}
```

Which when it is passed through the code you highlighted will result in returning `ctrl.Result{}, nil` which should be exactly the same thing we did in the original code.

The other place in the steps we do this is:
```golang
func (r *ClowdEnvironmentReconciliation) setToBeDisabled() (ctrl.Result, error) {
	if r.env.Spec.Disabled {
		r.log.Info("Reconciliation aborted - set to be disabled")
		return ctrl.Result{}, errors.New(SKIPRECONCILE)
	}
	return ctrl.Result{}, nil
}
```

Which was based on and should be logically equivilent to:
```golang
	if app.Spec.Disabled {
		log.Info("Reconciliation aborted - set to be disabled")
		return ctrl.Result{}, nil
	}
```

So, from the way I parse all this we should be doing exactly what the old implementation was doing. Am I parsing this out wrong? If not, then is this a problem that already existed in Clowder? And if so, what so I do in these cases?

Comment (psav) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]:
```
func (r *ClowdAppReconciliation) isEnvLocked() (ctrl.Result, error) {
	if ReadEnv() == r.app.Spec.EnvName {
		r.recorder.Eventf(r.app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", r.app.Spec.EnvName)
		r.log.Info("Env currently being reconciled")
		return ctrl.Result{Requeue: true}, errors.New(SKIPRECONCILE)
	}
	return ctrl.Result{}, nil
}
```

I think this function is the one that breaks. It says the Clowder Env is Locked, and it should be requeued. But it we squash the error and return `nil`, this won't be the case.

Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]: OK, I changed it but that's what we did before this change:

```golang
	if ReadEnv() == app.Spec.EnvName {
		r.Recorder.Eventf(&app, "Warning", "ClowdEnvLocked", "Clowder Environment [%s] is locked", app.Spec.EnvName)
		log.Info("Env currently being reconciled")
		return ctrl.Result{Requeue: true}, nil
	}
}
```

Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]: We're failing on merge conflict again. There were more changes to the deployment implementation so we're dying on that. I tried to resolve it but I'm not sure what changed - the diff output is a mess again due to how big the delta is in this branch.

Comment (psav) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]: I may have a large apology to make

Comment (psav) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]: I think I had forgotten the state we had where this was wrong
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler

Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go:150]: Cool cool - flipped it back around
